### PR TITLE
Add Argument to Specify YAML File Path

### PR DESCRIPTION
**Pull request by rudra-iitm**
This update introduces a new argument to allow users to specify the file path to the YAML file (snapcraft.yaml or rockcraft.yaml). It ensures the workflow operates seamlessly even when the YAML file is located in a non-conventional directory.

**Key Change**
- Added a `--yaml-path` argument to accept the path to the YAML file.

**Usage**
Users can now pass the file path using the `--yaml-path` flag: `--yaml-path: 'path/to/yaml/file'`

**Benefits**
- Enables workflows to handle diverse project structures.
- Improves usability for projects with custom file layouts.

### CODE REVIEW

**Comment by rudra-iitm** [comment]
@sergio-costas , please review this PR. The unittests are passing successfully; however, the style check is failing for a file that I did not modify. The error states "too-many-positional-arguments", even though Pylint is explicitly disabled for that line. I’m unable to determine the cause of this failure.

**Comment by sergio-costas** [comment]
@rudra-iitm I'll check it.

**Comment by sergio-costas** [comment]
Mmm... odd... in local it does work...

Maybe updating the ubuntu version to ubuntu-noble (ubuntu-latest isn't "the latest one", but... 18.04 I think) and python version to 3.12 in the .github/workflows/updatesnaptests.yml file...

**Comment by sergio-costas** [comment]
(I don't want to send a patch for that because every time that I tried to send a commit to a PR that wasn't mine, I broke the git history. So I prefer to stay in the safe side 😅 )

**Comment by rudra-iitm** [comment]
> Mmm... odd... in local it does work...
> 
> Maybe updating the ubuntu version to ubuntu-noble (ubuntu-latest isn't "the latest one", but... 18.04 I think) and python version to 3.12 in the .github/workflows/updatesnaptests.yml file...

I am working on a Mac and using Python 3.12. The style check is still failing locally for me.

**Comment by rudra-iitm** [comment]
> (I don't want to send a patch for that because every time that I tried to send a commit to a PR that wasn't mine, I broke the git history. So I prefer to stay in the safe side 😅 )

😅 Better safe than sorry

**Comment by sergio-costas** [comment]
Mmm... which version of pylint and flake8 do you have? Mine are:

```
raster ~/workspace/desktop-snaps/updatesnap (stable)$ pylint --version
pylint 3.2.7
astroid 3.2.2
Python 3.12.7 (main, Nov  6 2024, 18:29:01) [GCC 14.2.0]
raster ~/workspace/desktop-snaps/updatesnap (stable)$ flake8 --version
7.1.1 (mccabe: 0.7.0, pycodestyle: 2.11.1, pyflakes: 3.2.0) CPython 3.12.7 on Linux
```

**Comment by sergio-costas** [comment]
I found how to fix it. Please, remove your changes in the workflows and rebase against `stable`. It was a problem with the pylint version.

**Comment by sergio-costas** [comment]
@rudra-iitm I cancelled the test because I was wrong: it's not "ubuntu-noble" but "ubuntu-24.04". Anyway, in the main branch is everything fixed, so just removing that last commit and rebasing should be enough.

**Comment by rudra-iitm** [comment]
> @rudra-iitm I cancelled the test because I was wrong: it's not "ubuntu-noble" but "ubuntu-24.04". Anyway, in the main branch is everything fixed, so just removing that last commit and rebasing should be enough.

Done rebasing

**Comment by sergio-costas** [comment]
And checks have passed :-)

File: updatesnap/updatesnapyaml.py [hunk 1]
```diff
@@ -54,12 +53,6 @@ def get_yaml_file(self, project_url):
                 data = self._github.get_file(project_url, yaml_path)
             except (ValueError, ConnectionError):
                 data = None
-        if not data:
```
Comment (sergio-costas) [inline updatesnap/updatesnapyaml.py]: Why are you removing this? I think that it was needed for some case...

**Comment by rudra-iitm** [comment]
> And checks have passed :-)

Could you please review the changes and merge the PR?

File: updatesnap/updatesnapyaml.py [hunk 2]
```diff
@@ -40,10 +40,9 @@ def get_working_branch(self, project_url):
                 break
         return working_branch
 
-    def get_yaml_file(self, project_url):
+    def get_yaml_file(self, project_url, yaml_path):
         """ Searches in a project for the 'snapcraft.yaml' file and
```
Comment (sergio-costas) [inline updatesnap/updatesnapyaml.py:44]: I dislike how this has been implemented. This presumes that `snapcraft.yaml` will be passed by default in the `yaml_path` parameter, and if another path is passed, it still will try to search in `snap/snapcraft.yaml` if that path fails for whatever reason. I think that the default value in `args` should be `None`, and in this function, if the `yaml_path` value is `None`, should search in the common places using the old code, and if it isn't `None`, should search only in the passed path and fail if it is an invalid file or it doesn't exist.

**Review by sergio-costas — CHANGES_REQUESTED** [review]
Some details.

**Comment by sergio-costas** [comment]
Yes, sorry... I reviewed it yesterday, but forgot to press the button to commit it.

File: updatesnap/updatesnapyaml.py [hunk 3]
```diff
@@ -76,6 +83,8 @@ def main():
                         help='Version schema of snapping repository')
     parser.add_argument('--rock-version-schema', action='store', default='None',
                         help='Version schema of rock repository')
+    parser.add_argument('--yaml-path', action='store', default='None',
```
Comment (sergio-costas) [inline updatesnap/updatesnapyaml.py]: I think that it would be better to use `None` instead of the string `"None"`. Like in `github-user` and `github-token`.

Also, I see that `version-schema` and `rock-version-schema` also use a string instead of the `None` native type... I think that it should be fixed in another PR.

File: updatesnap/updatesnapyaml.py [hunk 4]
```diff
@@ -40,9 +40,16 @@ def get_working_branch(self, project_url):
                 break
         return working_branch
 
-    def get_yaml_file(self, project_url):
+    def get_yaml_file(self, project_url, yaml_path):
         """ Searches in a project for the 'snapcraft.yaml' file and
             returns its contents """
+        if yaml_path != 'None':
```
Comment (sergio-costas) [inline updatesnap/updatesnapyaml.py]: And, of course, this should be changed to `if yaml_path is not None:` after doing the change in the default value.

**Review by sergio-costas — CHANGES_REQUESTED** [review]
Some nitpicks.

File: updatesnap/updatesnapyaml.py (hunk 4 above)
Comment (rudra-iitm) [reply updatesnap/updatesnapyaml.py]: Done refracting code

File: updatesnap/updatesnapyaml.py (hunk 3 above)
Comment (rudra-iitm) [reply updatesnap/updatesnapyaml.py]: I will do it in new PR for version-schema and rock-version-schema.

**Comment by rudra-iitm** [comment]
@sergio-costas I often forget to update action.yml because everything works fine during local testing without it. Is there a way to test the workflow directly (instead of locally) to ensure I don’t miss updating action.yml?

File: updatesnap/updatesnapyaml.py (hunk 3 above)
Comment (rudra-iitm) [reply updatesnap/updatesnapyaml.py]: I think there is an issue when using None instead of 'None', as action.yml sends 'None' when an argument is not passed. We can't pass None as the default value in action.yml because it breaks.

Comment (sergio-costas) [reply updatesnap/updatesnapyaml.py]: Maybe you should try a different way in the `action.yml` file. You can test if an environment variable is defined or not directly in the line, and add a piece of text if it is defined. For example:

    ${VERSION_SCHEMA:+--version-schema $VERSION_SCHEMA}

the `:+` part means "if the previous variable is defined, add what is after me". Thus, if VERSION_SCHEMA isn't defined, that will translate to no text, while if it is defined, it will translate to `--version-schema $VERSION_SCHEMA`

**Comment by sergio-costas** [comment]
About testing the action.yml... there seems to be "gh", the github-cli program. I think that it allows to do that.

**Comment by rudra-iitm** [comment]
Refactored `action.yml` to handle undefined `yaml-path` arg as well. Please review the changes once again.

**Review by sergio-costas — APPROVED** [review]
After the changes, it looks good.

**Comment by sergio-costas** [comment]
@rudra-iitm Do you have merge permissions...?

**Comment by rudra-iitm** [comment]
> @rudra-iitm Do you have merge permissions...?

No, I don't have such permissions.

**Comment by sergio-costas** [comment]
I'll merge it, then. And thanks for the PR!
